dashboard/modules/log/log_agent.py

```diff
@@ -16,6 +18,208 @@
 logger = logging.getLogger(__name__)
 routes = dashboard_optional_utils.ClassMethodRouteTable
 
+# 64 KB
+BLOCK_SIZE = 1 << 16
+
+# Keep-alive interval for reading the file
+DEFAULT_KEEP_ALIVE_INTERVAL_SEC = 1
+
+
+def _find_tail_start_from_last_lines(
```

Comment:
Q: Why can't we jsut modify the existing function? Isn't it that we just need to modify existing func this way?

```python3
1. Instead of 0, we allow to pass end_offset
    # Seek to the end of the file
    f.seek(END_OFFSET, 2)
    block_end_byte = f.tell()

    last_byte_read = block_end_byte
    lines_to_go = total_lines_wanted
    block_number = -1
    blocks = []
```
```python3
    # Read blocks into memory until we have seen at least
    # `total_lines_wanted` number of lines. Then, return a string
    # containing the last `total_lines_wanted` number of lines
    while lines_to_go > 0 and block_end_byte > 0:
        if block_end_byte - BLOCK_SIZE > 0:
            f.seek(block_number * BLOCK_SIZE, 2)
            blocks.append(f.read(BLOCK_SIZE))
        else:
            f.seek(0, 0)
            blocks.append(f.read(block_end_byte))
        lines_found = blocks[-1].count(b"\n")
        lines_to_go -= lines_found
        block_end_byte -= BLOCK_SIZE
        block_number -= 1
2. Here, we can just call .tail to get the start_offset? 
    all_read_text = b"".join(reversed(blocks))
    return b"\n".join(all_read_text.splitlines()[-total_lines_wanted:]), last_byte_read
```

Reply:
Synced offline - it's a bit trickier if we want to 
1. not store line content in memory 
2. handle block read boundary properly.